runtime/lua/vim/lsp/semantic_tokens.lua

```diff
@@ -0,0 +1,142 @@
+---   }
+--- </pre>
+---
+--- `on_invalidate_range`: A function with signature `function(ctx, line_start, line_end)` called whenever tokens
+---                        in a specific line range (`line_start`, `line_end`) should be considered invalidated
+---                        (see |lsp-handler| for the definition of `ctx`). `line_end` can be -1 to
+---                        indicate invalidation until the end of the buffer.
+function M.on_full(err, response, ctx, config)
+  if not ctx.bufnr or not ctx.client_id then
+    return
+  end
+  active_requests[ctx.bufnr] = false
+  if config and config.on_invalidate_range then
+    config.on_invalidate_range(ctx, 0, -1)
+  end
+  -- if tick has changed our response is outdated!
+  -- FIXME: this is should be done properly here and in the codelens implementation. Handlers should
+  -- not be responsible of checking whether their responses are still valid.
+  if err or not response or not config.on_token or last_tick[ctx.bufnr] ~= vim.api.nvim_buf_get_changedtick(ctx.bufnr) then
+    return
+  end
+  local client = vim.lsp.get_client_by_id(ctx.client_id)
+  if not client then
+    return
+  end
+  local legend = client.server_capabilities.semanticTokensProvider.legend
+  local token_types = legend.tokenTypes
+  local token_modifiers = legend.tokenModifiers
+  local data = response.data
+
+  local line
+  local start_char = 0
+  for i = 1, #data, 5 do
+    local delta_line = data[i]
+    line = line and line + delta_line or delta_line
+    local delta_start = data[i + 1]
+    start_char = delta_line == 0 and start_char + delta_start or delta_start
+
+    -- data[i+3] +1 because Lua tables are 1-indexed
+    local token_type = token_types[data[i + 3] + 1]
+    local modifiers = modifiers_from_number(data[i + 4], token_modifiers)
+
+    local token = {
+      line = line,
+      start_char = start_char,
+      length = data[i + 2],
+      type = token_type,
+      modifiers = modifiers,
+      offset_encoding = client.offset_encoding,
+    }
+
+    if token_type and config and config.on_token then
+      config.on_token(ctx, token)
+    end
+  end
+end
+
+--- |lsp-handler| for the method `textDocument/semanticTokens/refresh`
+---
+function M.on_refresh(err, _, ctx, _)
+  if not err then
+    for _, bufnr in ipairs(vim.lsp.get_buffers_by_client_id(ctx.client_id)) do
+      M.refresh(bufnr)
+    end
+  end
+  return vim.NIL
+end
+
+---@private
+function M._save_tick(bufnr)
+  bufnr = bufnr or vim.api.nvim_get_current_buf()
+  last_tick[bufnr] = vim.api.nvim_buf_get_changedtick(bufnr)
+  active_requests[bufnr] = true
+end
+
+--- Refresh the semantic tokens for the current buffer
+---
+--- It is recommended to trigger this using an autocmd or via keymap.
+---
+--- <pre>
+---   autocmd BufEnter,CursorHold,InsertLeave <buffer> lua require 'vim.lsp.semantic_tokens'.refresh()
+--- </pre>
+---
+--- @param bufnr number|nil
```

Comment:
In other places (diagnostic API for example) we use `nil` to mean "all buffers" and 0 for current. Maybe worth changing this here to either not accept `nil` at all or have the same semantics for consistency.

Reply:
sample implementation: https://github.com/theHamsta/nvim-semantic-tokens it was out-factored of this PR so that we don't need to fix how the mapping from the VSCode modifiers to hl-groups should work

Reply:
The intention was to make it work like https://github.com/theHamsta/neovim/blob/3b0a0c6ca6c5f4e719028be2dba11853ffec6b6b/runtime/lua/vim/lsp/codelens.lua#L231-L249

Reply:
I also have an implementation loosely based on @theHamsta's but makes it easier to define color schemes without relying on lua config. It's based entirely on which highlight groups are defined following a specific pattern (similar to coc.nvim's method).

I still need to pull it into a separate plugin in its own git repo but for now it's a [directory in my dotfiles](https://github.com/jdrouhard/dotfiles/tree/master/home/.config/nvim/lua/semantic-tokens)

Reply:
remove `nil`